File: Objective C/MoPub Mediation/ThumbnailExample/ThumbnailExample/ViewController.m
//
//  ViewController.m
//  BannerExample
//
//  Copyright © 2020 Ogury Co. All rights reserved.
//

#import "ViewController.h"
#import "BlackListViewController.h"
#import <OguryChoiceManager/OguryChoiceManager.h>
#import <MoPubSDK/MoPub.h>

@interface ViewController()<MPAdViewDelegate>

@property (nonatomic, weak) IBOutlet UITextView *statusTextView;
@property (nonatomic, strong) MPAdView *thumbnail;
@property (nonatomic, assign) BOOL isAdLoaded;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self addNewStatus: @"Choice Manager Loading..."];
    
    //The setup of Ogury Choice Manager and is done AppDelegate file.
    [[OguryChoiceManager sharedManager] askWithViewController:self andCompletionBlock:^(NSError * _Nullable error, OguryChoiceManagerAnswer answer) {
        if (!error) {
            switch (answer) {
                case OguryChoiceManagerAnswerNoAnswer:
                    [self addNewStatus: @"Choice Manager No Answer"];
                    break;
                case OguryChoiceManagerAnswerFullApproval: // TCF Option
                    [self addNewStatus: @"Choice Manager Full Approval"];
                    break;
                case OguryChoiceManagerAnswerPartialApproval: // TCF Option
                    [self addNewStatus: @"Choice Manager Partial Approval"];
                    break;
                case OguryChoiceManagerAnswerRefusal: // TCF Option
                    [self addNewStatus: @"Choice Manager Refusal"];
                    break;
                case OguryChoiceManagerAnswerSaleAllowed: // CCPA Option
                    [self addNewStatus: @"Choice Manager Sale Allowed"];
                    break;
                case OguryChoiceManagerAnswerSaleDenied: // CCPA Option
                    [self addNewStatus: @"Choice Manager Unknown Option"];
                    break;
            }
        } else {
            [self addNewStatus: [NSString stringWithFormat:@"Choice Manager error : %@", error.description]];
        }
    }];
}

- (IBAction)loadAdBtnPressed:(id)sender {
    [self addNewStatus: @"Thumbnail Loading ..."];
    self.thumbnail = [[MPAdView alloc] initWithAdUnitId:@"mopub_adunit"];
    self.thumbnail.delegate = self;
    self.thumbnail.maxAdSize = CGSizeMake(200, 200);
    [self.thumbnail stopAutomaticallyRefreshingContents];
    [self.thumbnail setLocalExtras:
     @{
         @"rect_corner": @"bottom_right",
         @"x_margin": @30,
         @"y_margin": @30,
         @"whitelist": @[@"com.example.bundle", @"com.example.bundle2"],
         @"blacklist": @[NSStringFromClass([BlackListViewController class])]
        
     }];
    [self.thumbnail loadAd];
}

- (IBAction)showAdBtnPressed:(id)sender {
    if (self.isAdLoaded == YES) {
        [self addNewStatus: @"Ad requested to show"];
        [self.view addSubview:self.thumbnail];
    }
}

- (void)addNewStatus:(NSString *)status {
    NSString * statusLog = [status stringByAppendingString:@"\n"];
    self.statusTextView.text = [self.statusTextView.text stringByAppendingString:statusLog];
    NSRange bottom = NSMakeRange(self.statusTextView.text.length-1, 1);
    [self.statusTextView scrollRangeToVisible:bottom];
}

#pragma mark - MoPub Delegate

- (UIViewController *)viewControllerForPresentingModalView {
    return self;
}

- (void)adViewDidLoadAd:(MPAdView *)view adSize:(CGSize)adSize {
    [self addNewStatus: @"Ad received"];
    self.isAdLoaded = YES;
}

- (void)adView:(MPAdView *)view didFailToLoadAdWithError:(NSError *)error {
    [self addNewStatus: [NSString stringWithFormat:@"Error %@", error.description]];
}

@end

File: Swift/MoPub Mediation/ThumbnailExample/ThumbnailExample/ViewController.swift
//
//  ViewController.swift
//  ThumbnailExample
//
//  Copyright © 2020 Ogury Co. All rights reserved.
//

import UIKit
import OguryChoiceManager
import MoPubSDK

class ViewController: UIViewController {
    
    @IBOutlet weak var statusTextView: UITextView!
    var thumbnail: MPAdView?
    var adLoaded: Bool = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        addNewStatus("Choice Manager Loading...")
        
        //The setup of Ogury Choice Manager is done AppDelegate.swift file.
        OguryChoiceManager.shared().ask(with: self) { (error, answer) in
            if error == nil {
                switch answer {
                case .noAnswer: // TCF Option
                    self.addNewStatus("Choice Manager No Answer")
                case .fullApproval: // TCF Option
                    self.addNewStatus("Choice Manager Full Approval")
                case .partialApproval: // TCF Option
                    self.addNewStatus("Choice Manager Partial Approval")
                case .refusal: // TCF Option
                    self.addNewStatus("Choice Manager Refusal")
                case .saleAllowed: // CCPA Option
                    self.addNewStatus("Choice Manager Sale Allowed")
                case .saleDenied: // CCPA Option
                    self.addNewStatus("Choice Manager Sale Denided")
                default:
                    self.addNewStatus("Choice Manager Unknown Option")
                }
            } else {
                self.addNewStatus("Choice Manager error : \(error.debugDescription)")
            }
        }
    }

    
    @IBAction func loadAdBtnPressed(_ sender: Any) {
        addNewStatus("Loading Ad...")
        thumbnail = MPAdView.init(adUnitId: "mopub_adunit")
        thumbnail?.delegate = self;
        thumbnail?.maxAdSize = CGSize(width: 200, height: 200)
        thumbnail?.stopAutomaticallyRefreshingContents()
        
        
        thumbnail?.localExtras = [
            "rect_corner": "bottom_right",
            "x_margin": 30,
            "y_margin": 30,
            "whitelist": ["com.example.bundle", "com.example.bundle2"],
            "blacklist": [BlackListViewController.className()]
        ];
        
        thumbnail?.loadAd()
    }
    
    @IBAction func showAdBtnPressed(_ sender: Any) {
        guard let thumbnailView = thumbnail, adLoaded else {
            addNewStatus("Ad not loaded")
            return
        }
        addNewStatus("Ad requested to show")
        self.view.addSubview(thumbnailView)
    }

    func addNewStatus(_ status: String) {
        DispatchQueue.main.async {
            let textToLog = status + "\n"
            self.statusTextView.textStorage.append(NSAttributedString(string: textToLog))
            let bottom = NSMakeRange(self.statusTextView.text.count - 1, 1)
            self.statusTextView.scrollRangeToVisible(bottom)
        }
    }
}

extension ViewController : MPAdViewDelegate {
    func viewControllerForPresentingModalView() -> UIViewController! {
        return self;
    }
    
    func adViewDidLoadAd(_ view: MPAdView!, adSize: CGSize) {
        addNewStatus("Ad received")
        self.adLoaded = true
    }
    
    func adView(_ view: MPAdView!, didFailToLoadAdWithError error: Error!) {
        addNewStatus("Error: \(error.debugDescription)")
    }
}

class BlackListViewController: UIViewController {
    override func viewDidLoad() {
        super.viewDidLoad();
    }
}
